### PR TITLE
Grid map unification

### PR DESCRIPTION
**Pull request by mishrasubhransu**
Used shared_ptr instead of pointers.
Used vectors<> instead of arrays
Merged grid and map into GridCore class and added more function
CarGeometry has a safety buffer around it
Added better visualization for car and primitives.
Templated CarConnectivity, simplified it as well.
CarCost handles all obstacles avoidance and gives consistent cost for true cost and heur cost.

---
This change is [Reviewable](https://reviewable.io/reviews/jhu-asco/dsl/5)

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -39,6 +42,9 @@ add_subdirectory (bin)
 INSTALL(FILES dsl.pc
         DESTINATION lib/pkgconfig)
 
+#Install the header only serialization library called cereal
+INSTALL(DIRECTORY third_party/cereal DESTINATION include)
```
Comment (marinkobi) [inline CMakeLists.txt]: I though we were going to use protobuf?

File: bin/cartest.cc [hunk 2]
```diff
@@ -21,32 +51,48 @@ vector<Vector2d> ToVector2dPath(const CarPath &path) {
...
-    cout << "\t\t output will be written to graphics files path1.ppm and path2.pppm" << endl;
+    cout << "\t\t output will be written to graphics files path1.ppm and path2.ppm" << endl;
     return 0;
   }
   assert(argc == 2);
 
   Params params(argv[1]);
 
+  //To plot the path as rectanges or just points
+  bool plot_car; params.GetBool("plot_car", plot_car);
```
Comment (marinkobi) [inline bin/cartest.cc]: split in two lines

File: bin/cartest.cc [hunk 3]
```diff
@@ -55,11 +101,15 @@ int main(int argc, char** argv)
   Vector3d gcs;
   params.GetVector3d("gcs", gcs);
 
-  Vector4d whxy;
-  bool useGeom = params.GetVector4d("geom", whxy);
-  
+  Vector5d whxy;
```
Comment (marinkobi) [inline bin/cartest.cc]: 5d?  rename variable to reflect this update

File: bin/cartest.cc [hunk 4]
```diff
@@ -55,11 +101,15 @@ int main(int argc, char** argv)
...
+  bool useGeom = params.GetVector5d("geom", whxy);
+  CarGeom geom;
+  if(useGeom)
+    geom.set(whxy(0),whxy(1),whxy(2),whxy(3),whxy(4));
+
   // load an occupancy map from ppm file
-  dsl::Map<bool, 2> omap = load(mapName.c_str(), ocs.tail<2>());
+  dsl::Map<bool, 2>::Ptr pomap = load(mapName, ocs.tail<2>());
+  dsl::Map<bool, 2>& omap = *pomap;
```
Comment (marinkobi) [inline bin/cartest.cc]: please use only one variable here, i.e. just omap as a shared pointer, othewise it's confusing

File: bin/cartest.cc [hunk 5]
```diff
@@ -69,87 +119,210 @@ int main(int argc, char** argv)
...
   if (!cmapValid) {
-    cmap = new dsl::Map<bool, 3>(xlb, xub, ocs);
+    cmap.reset(new dsl::Map<bool, 3>(xlb, xub, ocs));
     std::cout << "Making cmap... " << std::endl;
-
+    timer_start(&timer);
     if (useGeom) {
-      CarGeom geom(whxy(0), whxy(1), whxy(2), whxy(3));
-      CarGrid::MakeMap(geom, omap, *cmap);
+      int nthreads; params.GetInt("nthreads", nthreads);
```
Comment (marinkobi) [inline bin/cartest.cc]: splint into two lines

File: bin/cartest.cc [hunk 6]
```diff
@@ -69,87 +119,210 @@ int main(int argc, char** argv)
...
-  params.GetBool("onlyfwd", onlyfwd);
-  connectivity.SetPrimitives(dt, vx, kmax, kseg, onlyfwd);
-
-  cout << "Creating a graph..." << endl;
-  // create planner
-  struct timeval timer;
-  timer_start(&timer);
+  double ac;
+  Vector3d wt;
+  shared_ptr<CarCost> cost;
```
Comment (marinkobi) [inline bin/cartest.cc]: this doesn't need to be a shared pointer. You can just set it as an object passed by reference.

File: bin/cartest.cc [hunk 7]
```diff
@@ -69,87 +119,210 @@ int main(int argc, char** argv)
...
+
+    //Display the primitive at start
+    vector<vector<Vector2d>> prims(0);
+    bool gotprims = connectivity.GetPrims(start,prims);
+    if(gotprims){
+      saveMapWithPrims(dmap, "prim2.ppm",prims,3);
+      cout << "Map, with primitives from start saved to prim2.ppm" << endl;
+    }
+
+  }
```
Comment (marinkobi) [inline bin/cartest.cc]: these are very good additions! 
one thing though is that this file used to be much shorter and simpler, and now it's gotten too long. We should factorize it so that all the path plotting and the new stuff happens as part of the library not as part of this test. 
The tests ideally should be short and should not introduce a lot of new functionality

File: lib/carconnectivity.h [hunk 8]
```diff
@@ -52,13 +77,79 @@ class CarConnectivity : public GridConnectivity< Eigen::Vector3d, Eigen::Matrix3
...
+  }
+
+
+  /**
+   * Generates primitives based on the configuration
+   * @param dt time duration
+   * @param cfg the configuration for generation of primitives
+   * @return
+   */
+  bool SetPrimitives(double dt, CarPrimitiveConfig& cfg){
```
Comment (marinkobi) [inline lib/carconnectivity.h:103]: this should be in a *.cc file

File: lib/carconnectivity.h [hunk 9]
```diff
@@ -80,36 +171,91 @@ class CarConnectivity : public GridConnectivity< Eigen::Vector3d, Eigen::Matrix3
...
-   * @param v body fixed velocity (vx,vy,w)
-   * @return true on success, false if obststructed by obstacle
+   * Utility function to get all the primitives starting at a given position. Each primitive is a vector of
+   * positions along the path. This is particularly useful for plotting the primitives.
+   * @param prims A single point along a primitive is xy pos
    */
-  bool Flow(std::tuple< SE2Cell*, SE2Path, double>& pathTuple,
-            const Eigen::Matrix3d& g0,
-            const Eigen::Vector3d& v) const;
+  bool GetPrims(const Vector3d pos, vector<vector<Vector2d>>& prims ){
```
Comment (marinkobi) [inline lib/carconnectivity.h]: should not be in header

File: lib/carcost.cc [hunk 10]
```diff
@@ -2,23 +2,80 @@
...
+  const Matrix3d& ga = a.data;
+  const Matrix3d& gb = b.data;
+  Matrix3d gai; se2_inv(gai,ga);
+  Matrix3d dg = gai*gb;
+  Vector3d twist; se2_log(twist,dg);
 
-  return (1 - eps)*(d1 + ac * d2);
-}
+  if(use_twistnorm_metric){
+    return (1 - eps)*(twist.array()* wt.array()).matrix().norm();
```
Comment (marinkobi) [inline lib/carcost.cc]: that looks awkward, don't you just need an dot product?

File: lib/carcost.cc [hunk 11]
```diff
@@ -2,23 +2,80 @@
...
+    cout<<"wrong wt in carcost"<<endl;
+    this->wt << 0.1, 1, 2;
+  }
 }
 
-double CarCost::Real(const SE2Cell& a, const SE2Cell& b) const {
-  double d1 = (a.c.tail< 2 >() - b.c.tail< 2 >()).norm(); // position distance
-  double d2 = fabs(fangle(a.c(0) - b.c(0))); // orientation distance
-  return d1 + ac * d2;
+double CarCost::Real(const SE2Cell& a, const SE2Cell& b) const{
```
Comment (marinkobi) [inline lib/carcost.cc]: this function looks expensive, where is it used?  
the real cost ideally should be computed incrementally at the same time when one integrates the edges (e.g. inside Connectivity)

File: lib/cargeom.cc [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+void CarGeom::Raster(const Eigen::Vector2d &cs, std::vector<Eigen::Vector2d> &points) const {
+  points.clear();
+  for (double x = -le_/2; x < le_/2; x += cs[0])
+    for (double y = -be_/2; y < be_/2; y += cs[1])
+      points.push_back(Eigen::Vector2d(x + ox_, y + oy_));
+}
+
+double CarGeom::l() const {return l_;}
```
Comment (marinkobi) [inline lib/cargeom.cc]: it's ok to define these inline in the header

File: lib/cargrid.cc [hunk 13]
```diff
@@ -11,198 +11,30 @@
...
-      assert(id3 < cmap.nc);
-      map.cells[id2] = cmap.cells[id3];
+//SE2CellGrid::Vectornb(true,false,false) says which dimensions are wrapped
+CarGrid::CarGrid(const Map<bool, 3> &cmap, const Vector3d& cs)
+    : SE2CellGrid(cmap.xlb, cmap.xub, cs, SE2CellGrid::Vectornb(true,false,false)){
+
+  //Iterate over all cells
+  auto fun = [&](int id, const Vector3i& gidx){
+    Vector3d cc; //CellCenter
+    bool gotcenter = CellCenter(cc,gidx);assert(gotcenter);
```
Comment (marinkobi) [inline lib/cargrid.cc]: split lines

File: lib/cargrid.h [hunk 14]
```diff
@@ -13,34 +13,13 @@
 #include <Eigen/Dense>
 #include <Eigen/Geometry>
 #include <vector>
-#include "map.h"
 
 namespace dsl {
+using std::shared_ptr;
```
Comment (marinkobi) [inline lib/cargrid.h]: not a good idea to include, just use std:: inline

File: lib/cell.h [hunk 15]
```diff
@@ -16,8 +16,10 @@ namespace dsl {
  * The cell can also contain a generic data of type T, which by default is just
  * a point in the cell.
  */
-template < class PointType, class DataType >
+template < class PT, class DT >
```
Comment (marinkobi) [inline lib/cell.h]: that's confusing, please revert to old syntax

File: lib/cargrid.h [hunk 16]
```diff
@@ -13,34 +13,13 @@
...
-    for (double x = 0; x < l; x += cs[0])
-      for (double y = 0; y < b; y += cs[1])
-        points.push_back(Eigen::Vector2d(x + ox, y + oy));
-  }
-
-  /// center of bounding rectangle
-};
-
+using SE2Cell =  Cell< Eigen::Vector3d, Eigen::Matrix3d >; // a cell that stores an SE(2) transformation  matrix
+using SE2CellPtr = shared_ptr<SE2Cell>;
```
Comment (marinkobi) [inline lib/cargrid.h]: using shared_ptr for cells is a bad idea... they are very expensive and should only be used for large objects that truly require shared access and have no clearly defined notion of who owns them

File: lib/grid.h [hunk 17]
```diff
@@ -10,178 +10,941 @@
...
+#include <fstream>
+
+#include "cereal/archives/binary.hpp" //serialization support
+#include "cereal/types/vector.hpp" //type support
+#include "cereal/types/eigen.hpp" //type support
 
 namespace dsl {
 
 struct EmptyData {};
+using std::shared_ptr;
```
Comment (marinkobi) [inline lib/grid.h]: we should not be including "using std::"

File: lib/grid.h [hunk 18]
```diff
@@ -10,178 +10,941 @@
...
+ *
+ * TODO: 1. Make sure origin lies at center of a cell
+ *       2. Check for bound in GetSlice and GetStack methods
+ *       3. Implement GetStack method that can perform occupancy dilation
+ *       4. Add clone methods to deep copy data
+ *
  */
-template < class PointType, class DataType = EmptyData>
- struct Grid {
+template < class PointType, class CellContent>
```
Comment (marinkobi) [inline lib/grid.h]: For consistency one should use Type for the second argument, why change the naming here?

File: lib/grid.h [hunk 19]
```diff
@@ -10,178 +10,941 @@
...
+    return xlb[i] + (idx +0.5)*cs[i];
+  }
+
+  /**
+   * Get id of point x useful for direct lookup in the cell array
+   * @param x point
+   * @return the cell array id. -1 if point x is not in grid
+   */
+  int Id(const Vectornd& x) const {
+    Vectorni gidx; Index(gidx,x);
```
Comment (marinkobi) [inline lib/grid.h]: split lines

File: lib/grid.h [hunk 20]
```diff
@@ -10,178 +10,941 @@
...
-    if (n==3)
-      return Index(x, 0) + gs[0]*Index(x, 1) + gs[0]*gs[1]*Index(x, 2);
+  /**
+   * Get the id of a point corresponding to the grid index
+   * @param gidx grid index
+   * @return the cell array id. -1 if gidx is not within bounds
+   */
+  int Id(const Vectorni& gidx) const {
+    if(Valid(gidx))
+      return gidx.transpose()*cgs;
```
Comment (marinkobi) [inline lib/grid.h]: the goal  was to make this more efficient by explicitly avoiding the transpose and multiplication for low-dimensional cases, why is it removed?

File: lib/grid.h [hunk 21]
```diff
@@ -10,178 +10,941 @@
...
-        cum *= gs[i];
+  /**
+   * Get the grid index of i coordinate/dimension
+   * @param xi value of a point at i coordinate/dimension
+   * @param i coordinated index/ dimension
+   * @return gidx[i], i.e. grid index along i coordinate/dimension. Meaningful even when x is out of bounds.
+   */
+  int Index(double xi, int i) const {
+    if(wd[i]){ //dimension is wrapped
+      while(xi < xlb[i]){xi += ds[i];}
```
Comment (marinkobi) [inline lib/grid.h]: using while here looks dangerous, is the loop guaranteed to execute at most twice? If not one should use mod()

File: lib/grid.h [hunk 22]
```diff
@@ -10,178 +10,941 @@
...
-   * @param i coordinate index
-   * @return index into cell array
    */
-  int Index(const PointType& x, int i) const {
-    return floor((x[i] - xlb[i]) / (xub[i] - xlb[i]) * gs[i]);
+  void Index(Vectorni& gidx, const Vectornd& x) const {
+    for(size_t i=0;i<n;i++){
+      double xi = x[i];
+      if(wd[i]){ //dimension is wrapped
+        while(xi < xlb[i]){xi += ds[i];}
```
Comment (marinkobi) [inline lib/grid.h]: same as above

File: lib/grid.h [hunk 23]
```diff
@@ -10,178 +10,941 @@
...
+                  id++;
+                }
+              }
+            }
+          }
+        }
+        break;
+      default:
+        Vectornp1i gidxe = Vectornp1i::Zero();//e stands for extra element
+        Vectornp1i  gse; gse << gs,0; //0 to indicate iteration over all cells over
```
Comment (marinkobi) [inline lib/grid.h]: split lines

File: lib/grid.h [hunk 24]
```diff
@@ -10,178 +10,941 @@
...
+      grid_coord[i] = (metric_coord[i] - point0).array()/cs.array();
+  }
+
+  /**
+   * Converts a set of points in metric coordinate to grid coordinates in placeS
+   * @param coord In metric coordinates. After update they are in grid coordinates
+   */
+  void ToGridCoordinates(vector<Vectornd>& coord) const{
+
+    Vectornd point0; CellCenter(point0,Vectorni::Zero());
```
Comment (marinkobi) [inline lib/grid.h]: split

File: lib/grid.h [hunk 25]
```diff
@@ -10,178 +10,941 @@
...
+ * containing data of type DataType.
+ * A grid provides instant access to the elements by maintaining
+ * an n-dimensional array of pointers to cells. Cells that are empty,
+ * e.g. that are inside obstacles, correspond to null pointers.
+ *
+ * Note that this data structure is only viable up to a few dimensions,
+ * e.g. dim=5 or 6.
+ */
+template < class PointType, class DataType = EmptyData>
+using Grid = GridCore< PointType,shared_ptr< Cell<PointType, DataType> > >;
```
Comment (marinkobi) [inline lib/grid.h]: not a good idea to use shared_ptr here

File: lib/gridsearch.h [hunk 26]
```diff
@@ -60,9 +60,11 @@ using std::vector;
  * there is cost of transitioning between two cells.
  *
  */
-template < class PointType, class DataType, class ConnectionType = std::vector< Cell<PointType, DataType>* > >
-  class GridSearch : public Search< Cell<PointType, DataType>, ConnectionType > {
+template < class PT, class DT, class ConnectionType = std::vector< Cell<PT, DT>* > >
+  class GridSearch : public Search< Cell<PT, DT>, ConnectionType > {
```
Comment (marinkobi) [inline lib/gridsearch.h]: please revert to the old names, these are not descriptive at all

File: lib/gridutils.cc [hunk 27]
```diff
@@ -0,0 +1,398 @@
...
+        int id3 = i + cmap.gs[0]*j + cmap.gs[0]*cmap.gs[1]*k; // 3d index into cmap
+        assert(id3 < cmap.nc);
+        cmap.cells[id3] = omap.cells[id2];
+      }
+    }
+  }
+  return false;
+}
+
+//bool MakeSE2Map(const CarGeom& geom, const Map<bool, 2> &omap, Map<bool, 3> &cmap, int nthreads) {
```
Comment (marinkobi) [inline lib/gridutils.cc]: remove if not needed

File: lib/grid.h [hunk 28]
```diff
@@ -10,178 +10,941 @@
...
+ * TODO: 1. Make sure origin lies at center of a cell
+ *       2. Check for bound in GetSlice and GetStack methods
+ *       3. Implement GetStack method that can perform occupancy dilation
+ *       4. Add clone methods to deep copy data
+ *
  */
-template < class PointType, class DataType = EmptyData>
- struct Grid {
+template < class PointType, class CellContent>
+class GridCore {
```
Comment (marinkobi) [inline lib/grid.h]: Why change the name to GridCore? Isn't this just a Grid? 
using Grid = GridCore< PointType,shared_ptr< Cell<PointType, DataType> > >; 
is confusing since Grid according to it becomes an alias for GridCore (that's contrary to thinking of Core as something residing inside the Grid).  
I would just call GridCore -> Grid
and then use a more specific alias, e.g.
SE2Grid  = Grid< Eigen::Vector3d, SE2Cell::Ptr>
or something like that -- i.e. all names should be consistent with the template arguments

Again, super important -- ::Ptr here should never be a shared pointer, consider using unique pointers instead

**Comment by marinkobi** [comment]
That's a lot of good functionality! 
We need to be a bit more careful in making sure that these new choices do not effect efficiency (see specific comments above). Also, the executable tests should be as simple as possible. They also serve as an illustration how to setup and run a problem, and if they're too long and complex it's hard to follow the logic. So it's good to have e.g. a super-simple cartest.cc and a more complex e.g. carviztest.cc with fancy visualization, but better yet factor out the plotting/visualization and call it with a line or two of code in all tests. 
Also, any chance we can use protobuf instead of cereal?

**Comment by marinkobi** [comment]
Also, we haven't been too strict about this, but any new functionality should follow
https://google.github.io/styleguide/cppguide.html
, since it looks like you have decided to follow some of the google-style approach, but not all of it, so we should be more consistent and just enforce it from now on. 

Also, I'm not sure that using vectors instead of arrays is a good idea, have you done benchmarking on it? The goal of using an array is because we know its constant size in advance and so it can be pre-allocated. Vectors ops can be more expensive in general.

**Comment by mishrasubhransu** [comment]
I made a new branch(google_styleguide) with the changes that you suggested above. Uses protobuf, follows google styleguide(new stuff) and more. I simplified the test examples and added good explanation at the beginning. 

I used vectors because it made life a lot easier. I haven't benchmarked it but, everyone at [stackoverflow](http://stackoverflow.com/questions/381621/using-arrays-or-stdvectors-in-c-whats-the-performance-gap) seems to be strongly suggesting to use vectors instead of using new. My vectors are allocated once in the constructors and after that the size can't be changed. If you would like I can do some benchmarking to see the performance difference. 

Lastly, you said in the above comment "We need to be a bit more careful in making sure that these new choices do not effect efficiency (see specific comments above)". I was unable to find any specific comments.

File: lib/carconnectivity.h (hunk 8 above)
Comment (mishrasubhransu) [reply lib/carconnectivity.h:103]: This is a template class, so shouldn't the implementation stay in the .h file?

File: lib/carconnectivity.h (hunk 9 above)
Comment (mishrasubhransu) [reply lib/carconnectivity.h]: This is a template class, so shouldn't the implementation stay in the .h file?

File: lib/carcost.cc (hunk 11 above)
Comment (mishrasubhransu) [reply lib/carcost.cc]: It is indeed done incrementally and the Real() is only called from inside Connectivity. It's basically what the connectivity used to do. It made sense to have both the costs together to make sure that the heuristics is admissible.

File: lib/carcost.cc (hunk 10 above)
Comment (mishrasubhransu) [reply lib/carcost.cc]: I need the a norm(scaled by the weight wt) of the twist, so dot product doesn't work.

File: lib/cell.h (hunk 15 above)
Comment (mishrasubhransu) [reply lib/cell.h]: I do it because I typedef them again as PointType and DataType so that it is accessible for use elsewhere like so
`using CarTwistConnectivity     = SE2GridConnectivity<SE2Cell::PointType, SE2Cell::DataType, SE2Twist>;`
`using CarPathConnectivity      = SE2GridConnectivity<SE2Cell::PointType, SE2Cell::DataType, SE2Path>;`

What do you suggest I do?

File: lib/grid.h (hunk 28 above)
Comment (mishrasubhransu) [reply lib/grid.h]: There were two different classes Grid and Map before, instantiated differently e.g. Grid<Vector3d, Matrix3d> and Map<3,bool> . While the cells in Grid stored Cell<Vector3d, Matrix3d>* , the cells in Map stored bools. 

To retain the same structure, I create this GridCore and aliased Grid and Map from it. Please advice how should  I structure them both finally.

File: lib/grid.h (hunk 20 above)
Comment (mishrasubhransu) [reply lib/grid.h]: I benchmarked against the old code(with checks for out of bound) and found the following elapsed times for 1e7 calls:

grid2.gs():10 10
dt Id:0.113313 secs
dt IdOld:0.0857138 secs

grid3.gs():10 10 10
dt Id:0.170228 secs
dt IdOld:0.127534 secs

grid4.gs():10 10 10 10
dt Id:0.220866 secs
dt IdOld:0.16996 secs

grid5.gs():10 10 10 10 10
dt Id:0.230863 secs
dt IdOld:0.212158 secs

Also gidx.transpose()*cgs is faster than gidx.dot(cgs)

File: lib/carconnectivity.h (hunk 8 above)
Comment (marinkobi) [reply lib/carconnectivity.h:103]: OK, I hand't noticed that it's now a template class -- we should use templates for sparingly, i.e. only for the most generic implementation -- carconnectivity is too specific and definitely should not be a templated class -- can we remove the dependence on ConnectionType and implement it differently?

File: lib/cell.h (hunk 15 above)
Comment (marinkobi) [reply lib/cell.h]: how about SE2Cell::Point ?

File: lib/grid.h (hunk 20 above)
Comment (marinkobi) [reply lib/grid.h]: yes, I would say 30% gain in efficiency is significant

File: lib/carconnectivity.h (hunk 8 above)
Comment (mishrasubhransu) [reply lib/carconnectivity.h:103]: I have different connection types and different data type, for which most of the code is identical.  What would be the best way to implement is without code duplication?
